src/storage/src/source/reclock.rs

```diff
@@ -8,413 +8,668 @@
-                    // Importantly, `continue`ing here is _correct_.  Just likely not ideal beacause we'll stall until
-                    // `self.now` catches up.
-                    error!("Time unexpectedly didn't go forwards for {:?}", self.name);
-                    tokio::time::sleep(self.update_interval).await;
-                    continue;
+        }
+    }
+
+    /// Ensures that the persist shard backing this reclock operator contains bindings that cover
+    /// the provided source frontier by minting bindings where appropriate.
+    ///
+    /// When this function returns the local dTVC view of the remap collection will contain
+    /// definite timestamp bindings that can be used to timestamp data at these offsets.
+    async fn mint<P: Borrow<PartitionId>>(
+        &mut self,
+        source_frontier: &HashMap<P, MzOffset>,
+    ) -> Result<(), Upper<Timestamp>> {
+        let mut updates = vec![];
+        for (pid, upper) in source_frontier {
+            let pid = pid.borrow();
+            let part_upper = self.last_source_upper.get(pid).cloned().unwrap_or_default();
+
+            if let Some(diff) = upper.checked_sub(part_upper) {
+                if diff > MzOffset::from(0) {
+                    updates.push((pid.clone(), diff));
                 }
-            };
-            let new_upper = Antichain::from_elem(new_ts + 1);
-
-            // Can happen if clock skew between replicas.
-            if PartialOrder::less_equal(&new_upper, &self.write_upper) {
-                info!(
-                    "Current write upper is not less than new upper for {:?}: {:?} {:?}",
-                    self.name, self.write_upper, new_upper
-                );
-                tokio::time::sleep(self.update_interval).await;
-                continue;
             }
-            let compare_and_append_result = self
-                .write_handle
-                .compare_and_append(
-                    new_bindings
-                        .iter()
-                        .map(|(partition, diff)| ((&(), *partition), &new_ts, diff)),
-                    self.write_upper.clone(),
-                    new_upper.clone(),
-                )
-                .await
-                .expect("Timestamper CAA")
-                .expect("Timestamper CAA 2");
-
-            // We don't just return here on success because the logic to determine what progress to return is complex
-            // and should be in one place only.
-            self.write_upper = match compare_and_append_result {
-                Ok(()) => new_upper,
-                Err(Upper(actual_upper)) => actual_upper,
-            };
+        }
 
-            while PartialOrder::less_than(&self.read_progress, &self.write_upper) {
+        // If there no updates we might still want to bump the upper time if enough time has passed
+        let mint_timestamp = if updates.is_empty() {
+            self.next_mint_timestamp().now_or_never()
```

Comment:
`now_or_never` has confusing semantics, I would just make a separate ts function that doesnt do the sleeping